src/mgr/Gil.h

```diff
@@ -0,0 +1,88 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LLC
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef GIL_H_
+#define GIL_H_
+
+#include "Python.h"
+
+#include "common/debug.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mgr
+#undef dout_prefix
+#define dout_prefix *_dout << "mgr " << __func__ << " "
+
+//
+// Use one of these in any scope in which you need to hold Python's
+// Global Interpreter Lock.  Do *not* nest these, as a second GIL
+// acquire will deadlock.  If in doubt, explicitly put a scope around
+// the block of code you know you need the GIL in.
+//
+// See the comment below for when to set new_thread == true
+//
+class Gil {
+public:
+  Gil(PyThreadState *ts, bool new_thread = false)
+  : pThreadState(ts), pNewThreadState(nullptr)
+  {
+    // Acquire the GIL, set the current thread state
+    PyEval_RestoreThread(pThreadState);
```

Comment:
If pThreadState is null it probably dies horribly. I've added an assert (so it'll still die horribly, but on our terms).

As for ensuring we haven't already been called, that's (sadly) up to the caller to be aware of. I've updated the comment to hopefully clarify this a bit more. If you invoke this ctor twice in the same thread it'll deadlock, e.g. do not do this:

````
void func1() {
  Gil gil(pMyThreadState);
  // Do some stuff that needs the GIL here

  func2();  // this will deadlock
}
void func2() {
  Gil gil(pMyThreadState);
  // Do some stuff that needs the GIL
}
````

You can avoid this deadlock like so:

````
void func1() {
  {  // extra scope
    Gil gil(pMyThreadState);
    // Do some stuff that needs the GIL here
  }

  func2();  // this won't deadlock
}
````

In summary, you have to be just as careful using this class as you need to be when using PyEvalRestoreThread(). AFAICT there's no public API for figuring out whether or not the GIL is already held.